File: warm_alarm_android/android/src/main/kotlin/com/andrew/alarm/WarmAlarmEventQueue.kt
package com.andrew.alarm

import android.content.Context
import android.content.SharedPreferences
import android.os.Build
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID

internal interface WarmAlarmEventQueueStorage {
    fun read(): String?

    fun write(value: String): Boolean
}

private class SharedPreferencesWarmAlarmEventQueueStorage(
    private val preferences: SharedPreferences,
) : WarmAlarmEventQueueStorage {
    override fun read(): String? = preferences.getString(KEY, null)

    override fun write(value: String): Boolean = preferences.edit().putString(KEY, value).commit()

    private companion object {
        const val KEY = "queue_v1"
    }
}

internal data class QueuedWarmAlarmEvent(
    val occurrenceId: String,
    val event: WarmAlarmEventWire,
)

internal interface WarmAlarmEventQueueStoreProtocol {
    fun enqueue(event: WarmAlarmEventWire): Boolean

    fun peek(): QueuedWarmAlarmEvent?

    fun remove(occurrenceId: String): Boolean
}

/**
 * A persistent FIFO. When [capacity] is reached, enqueue drops the oldest occurrence.
 */
internal class WarmAlarmEventQueueStore(
    private val storage: WarmAlarmEventQueueStorage,
    private val capacity: Int = DEFAULT_CAPACITY,
    private val occurrenceId: () -> String = { UUID.randomUUID().toString() },
) : WarmAlarmEventQueueStoreProtocol {
    init {
        require(capacity > 0)
    }

    override fun enqueue(event: WarmAlarmEventWire): Boolean =
        synchronized(storageLock) {
            val events = readAndRepair().toMutableList()
            events += QueuedWarmAlarmEvent(occurrenceId(), event)
            while (events.size > capacity) {
                events.removeAt(0)
            }
            persist(events)
        }

    fun loadAll(): List<QueuedWarmAlarmEvent> = synchronized(storageLock) { readAndRepair() }

    override fun peek(): QueuedWarmAlarmEvent? = synchronized(storageLock) { readAndRepair().firstOrNull() }

    override fun remove(occurrenceId: String): Boolean =
        synchronized(storageLock) {
            val events = readAndRepair()
            val retained = events.filterNot { it.occurrenceId == occurrenceId }
            if (retained.size == events.size) return@synchronized true
            persist(retained)
        }

    private fun readAndRepair(): List<QueuedWarmAlarmEvent> {
        val raw = storage.read() ?: return emptyList()
        val root = runCatching { JSONObject(raw) }.getOrNull()
        if (root == null || root.optInt("version", -1) != STORAGE_VERSION || root.optJSONArray("events") == null) {
            persist(emptyList())
            return emptyList()
        }
        val array = root.getJSONArray("events")
        val valid =
            buildList {
                for (index in 0 until array.length()) {
                    runCatching { decode(array.optJSONObject(index)) }.getOrNull()?.let(::add)
                }
            }
        if (valid.size != array.length()) {
            persist(valid)
        }
        return valid
    }

    private fun persist(events: List<QueuedWarmAlarmEvent>): Boolean {
        val encoded = JSONArray()
        events.forEach { encoded.put(encode(it)) }
        return storage.write(
            JSONObject()
                .put("version", STORAGE_VERSION)
                .put("events", encoded)
                .toString(),
        )
    }

    private fun encode(queued: QueuedWarmAlarmEvent): JSONObject =
        JSONObject()
            .put("occurrenceId", queued.occurrenceId)
            .put("alarmId", queued.event.alarmId)
            .put(
                "type",
                queued.event.type.name
                    .lowercase(),
            ).put("occurredAtMillis", queued.event.occurredAtMillis)
            .putNullable("snoozeDurationMillis", queued.event.snoozeDurationMillis)
            .putNullable("payload", queued.event.payload)
            .putNullable(
                "failureCode",
                queued.event.failure
                    ?.code
                    ?.raw,
            ).putNullable("failureMessage", queued.event.failure?.message)

    private fun decode(value: JSONObject?): QueuedWarmAlarmEvent? {
        value ?: return null
        val occurrenceId = value.optString("occurrenceId").takeIf { it.isNotEmpty() } ?: return null
        if (!value.has("alarmId") || !value.has("occurredAtMillis")) return null
        val type =
            runCatching { WarmAlarmEventTypeWire.valueOf(value.optString("type").uppercase()) }
                .getOrNull() ?: return null
        val failure =
            if (value.isNull("failureCode")) {
                null
            } else {
                val code = WarmAlarmFailureCodeWire.ofRaw(value.optInt("failureCode", -1)) ?: return null
                WarmAlarmFailureWire(code = code, message = value.optNullableString("failureMessage"))
            }
        return QueuedWarmAlarmEvent(
            occurrenceId = occurrenceId,
            event =
                WarmAlarmEventWire(
                    alarmId = value.getLong("alarmId"),
                    type = type,
                    occurredAtMillis = value.getLong("occurredAtMillis"),
                    snoozeDurationMillis = value.optNullableLong("snoozeDurationMillis"),
                    failure = failure,
                    payload = value.optNullableString("payload"),
                ),
        )
    }

    companion object {
        const val DEFAULT_CAPACITY = 64
        private const val STORAGE_VERSION = 1
        private const val PREFERENCES_NAME = "warm_alarm_event_queue"
        private val storageLock = Any()

        fun create(context: Context): WarmAlarmEventQueueStore {
            val storageContext =
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    context.createDeviceProtectedStorageContext()
                } else {
                    context
                }
            val preferences = storageContext.getSharedPreferences(PREFERENCES_NAME, Context.MODE_PRIVATE)
            return WarmAlarmEventQueueStore(SharedPreferencesWarmAlarmEventQueueStorage(preferences))
        }
    }
}

internal class WarmAlarmEventQueue(
    private val store: WarmAlarmEventQueueStoreProtocol,
    private val emit: (WarmAlarmEventWire, (Result<Unit>) -> Unit) -> Unit,
) {
    private var isDraining = false
    private var drainRequestedWhileActive = false

    fun enqueue(event: WarmAlarmEventWire): Boolean {
        if (!store.enqueue(event)) return false
        drain()
        return true
    }

    fun drain() {
        val shouldStart =
            synchronized(this) {
                if (isDraining) {
                    drainRequestedWhileActive = true
                    return@synchronized false
                }
                isDraining = true
                true
            }
        if (!shouldStart) return
        emitNext()
    }

    private fun emitNext() {
        val pending =
            synchronized(this) {
                drainRequestedWhileActive = false
                store.peek() ?: run {
                    isDraining = false
                    return@synchronized null
                }
            } ?: return
        emit(pending.event) { result ->
            if (result.isFailure || !store.remove(pending.occurrenceId)) {
                retryRequestedDrainOrStop()
                return@emit
            }
            emitNext()
        }
    }

    private fun retryRequestedDrainOrStop() {
        val shouldRetry =
            synchronized(this) {
                if (!drainRequestedWhileActive) {
                    isDraining = false
                    return@synchronized false
                }
                drainRequestedWhileActive = false
                true
            }
        if (shouldRetry) emitNext()
    }
}

internal fun migratePendingSnoozeEvents(
    legacyStore: PendingSnoozeEventStore,
    queueStore: WarmAlarmEventQueueStore,
): Boolean {
    for (pending in legacyStore.loadAll()) {
        if (!queueStore.enqueue(pending.event)) return false
        if (!legacyStore.remove(pending)) return false
    }
    return true
}

private fun JSONObject.putNullable(
    key: String,
    value: Any?,
): JSONObject = put(key, value ?: JSONObject.NULL)

private fun JSONObject.optNullableString(key: String): String? = if (isNull(key)) null else optString(key).takeIf { has(key) }

private fun JSONObject.optNullableLong(key: String): Long? =
    if (isNull(key) || !has(key)) null else runCatching { getLong(key) }.getOrNull()

File: warm_alarm_ios/ios/warm_alarm_ios/Sources/warm_alarm_ios/WarmAlarmEventQueue.swift
import Foundation

protocol WarmAlarmEventQueueStorage: AnyObject {
    func read() -> Data?
    func write(_ data: Data) -> Bool
}

/// Atomically replaces the queue file before reporting success. This covers process termination after return, not power loss.
final class FileWarmAlarmEventQueueStorage: WarmAlarmEventQueueStorage, @unchecked Sendable {
    private static let fileName = "event_queue_v1.json"
    private let fileURL: URL
    private let fileManager: FileManager

    init(
        fileURL: URL,
        fileManager: FileManager = .default
    ) {
        self.fileURL = fileURL
        self.fileManager = fileManager
    }

    convenience init(
        fileManager: FileManager = .default,
        bundle: Bundle = .main
    ) {
        let applicationSupportDirectory = fileManager.urls(
            for: .applicationSupportDirectory,
            in: .userDomainMask
        ).first ?? URL(fileURLWithPath: NSHomeDirectory(), isDirectory: true)
            .appendingPathComponent("Library/Application Support", isDirectory: true)
        let namespace = bundle.bundleIdentifier ?? "dev.flutter.pigeon.warm_alarm.host"
        let fileURL = applicationSupportDirectory
            .appendingPathComponent(namespace, isDirectory: true)
            .appendingPathComponent("warm_alarm", isDirectory: true)
            .appendingPathComponent(Self.fileName)
        self.init(fileURL: fileURL, fileManager: fileManager)
    }

    func read() -> Data? {
        try? Data(contentsOf: fileURL)
    }

    func write(_ data: Data) -> Bool {
        do {
            let directory = fileURL.deletingLastPathComponent()
            try fileManager.createDirectory(at: directory, withIntermediateDirectories: true)
            try? fileManager.setAttributes(
                [.protectionKey: FileProtectionType.completeUntilFirstUserAuthentication],
                ofItemAtPath: directory.path
            )
            try data.write(to: fileURL, options: .atomic)
            try? fileManager.setAttributes(
                [.protectionKey: FileProtectionType.completeUntilFirstUserAuthentication],
                ofItemAtPath: fileURL.path
            )
            return true
        } catch {
            return false
        }
    }
}

struct QueuedWarmAlarmEvent {
    let occurrenceID: String
    let event: WarmAlarmEventWire
}

/// A persistent FIFO. At 64 entries, enqueue drops the oldest occurrence.
final class WarmAlarmEventQueueStore: @unchecked Sendable {
    static let defaultCapacity = 64

    private let storage: WarmAlarmEventQueueStorage
    private let capacity: Int
    private let occurrenceID: () -> String
    private static let storageLock = NSLock()

    init(
        storage: WarmAlarmEventQueueStorage = FileWarmAlarmEventQueueStorage(),
        capacity: Int = defaultCapacity,
        occurrenceID: @escaping () -> String = { UUID().uuidString }
    ) {
        precondition(capacity > 0)
        self.storage = storage
        self.capacity = capacity
        self.occurrenceID = occurrenceID
    }

    func enqueue(_ event: WarmAlarmEventWire) -> Bool {
        Self.storageLock.lock()
        defer { Self.storageLock.unlock() }
        var events = readAndRepair()
        events.append(QueuedWarmAlarmEvent(occurrenceID: occurrenceID(), event: event))
        if events.count > capacity {
            events.removeFirst(events.count - capacity)
        }
        return persist(events)
    }

    func loadAll() -> [QueuedWarmAlarmEvent] {
        Self.storageLock.lock()
        defer { Self.storageLock.unlock() }
        return readAndRepair()
    }

    func remove(occurrenceID: String) -> Bool {
        Self.storageLock.lock()
        defer { Self.storageLock.unlock() }
        let events = readAndRepair()
        let retained = events.filter { $0.occurrenceID != occurrenceID }
        guard retained.count != events.count else { return true }
        return persist(retained)
    }

    private func readAndRepair() -> [QueuedWarmAlarmEvent] {
        guard let data = storage.read() else { return [] }
        guard
            let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
            (root["version"] as? NSNumber)?.intValue == 1,
            let rawEvents = root["events"] as? [Any]
        else {
            _ = persist([])
            return []
        }

        let valid = rawEvents.compactMap { raw -> QueuedWarmAlarmEvent? in
            guard let dictionary = raw as? [String: Any] else { return nil }
            return decode(dictionary)
        }
        if valid.count != rawEvents.count {
            _ = persist(valid)
        }
        return valid
    }

    private func persist(_ events: [QueuedWarmAlarmEvent]) -> Bool {
        let root: [String: Any] = [
            "version": 1,
            "events": events.map(encode),
        ]
        guard let data = try? JSONSerialization.data(withJSONObject: root, options: [.sortedKeys]) else {
            return false
        }
        return storage.write(data)
    }

    private func encode(_ queued: QueuedWarmAlarmEvent) -> [String: Any] {
        var value: [String: Any] = [
            "occurrenceId": queued.occurrenceID,
            "alarmId": queued.event.alarmId,
            "type": eventTypeName(queued.event.type),
            "occurredAtMillis": queued.event.occurredAtMillis,
        ]
        value["snoozeDurationMillis"] = queued.event.snoozeDurationMillis.map { $0 as Any } ?? NSNull()
        value["payload"] = queued.event.payload.map { $0 as Any } ?? NSNull()
        value["failureCode"] = queued.event.failure.map { $0.code.rawValue as Any } ?? NSNull()
        value["failureMessage"] = queued.event.failure?.message.map { $0 as Any } ?? NSNull()
        return value
    }

    private func decode(_ value: [String: Any]) -> QueuedWarmAlarmEvent? {
        guard
            let occurrenceID = value["occurrenceId"] as? String,
            !occurrenceID.isEmpty,
            let alarmID = (value["alarmId"] as? NSNumber)?.int64Value,
            let occurredAtMillis = (value["occurredAtMillis"] as? NSNumber)?.int64Value,
            let type = decodeEventType(value["type"])
        else {
            return nil
        }

        let failure: WarmAlarmFailureWire?
        if value["failureCode"] is NSNull || value["failureCode"] == nil {
            failure = nil
        } else {
            guard
                let rawCode = (value["failureCode"] as? NSNumber)?.intValue,
                let code = WarmAlarmFailureCodeWire(rawValue: rawCode)
            else {
                return nil
            }
            failure = WarmAlarmFailureWire(
                code: code,
                message: value["failureMessage"] as? String
            )
        }

        return QueuedWarmAlarmEvent(
            occurrenceID: occurrenceID,
            event: WarmAlarmEventWire(
                alarmId: alarmID,
                type: type,
                occurredAtMillis: occurredAtMillis,
                snoozeDurationMillis: (value["snoozeDurationMillis"] as? NSNumber)?.int64Value,
                failure: failure,
                payload: value["payload"] as? String
            )
        )
    }

    private func eventTypeName(_ type: WarmAlarmEventTypeWire) -> String {
        switch type {
        case .scheduled: "scheduled"
        case .fired: "fired"
        case .stopped: "stopped"
        case .snoozed: "snoozed"
        case .failed: "failed"
        }
    }

    private func decodeEventType(_ raw: Any?) -> WarmAlarmEventTypeWire? {
        if let rawValue = (raw as? NSNumber)?.intValue {
            return WarmAlarmEventTypeWire(rawValue: rawValue)
        }
        guard let name = raw as? String else { return nil }
        switch name {
        case "scheduled": return .scheduled
        case "fired": return .fired
        case "stopped": return .stopped
        case "snoozed": return .snoozed
        case "failed": return .failed
        default: return nil
        }
    }
}

final class WarmAlarmEventQueue: @unchecked Sendable {
    private let store: WarmAlarmEventQueueStore
    private let emit: (WarmAlarmEventWire, @escaping (Bool) -> Void) -> Void
    private let lock = NSLock()
    private var isDraining = false

    init(
        store: WarmAlarmEventQueueStore = WarmAlarmEventQueueStore(),
        emit: @escaping (WarmAlarmEventWire, @escaping (Bool) -> Void) -> Void
    ) {
        self.store = store
        self.emit = emit
    }

    @discardableResult
    func enqueue(_ event: WarmAlarmEventWire) -> Bool {
        guard store.enqueue(event) else { return false }
        drain()
        return true
    }

    func drain() {
        lock.lock()
        guard !isDraining else {
            lock.unlock()
            return
        }
        isDraining = true
        lock.unlock()
        emitNext()
    }

    private func emitNext() {
        guard let pending = store.loadAll().first else {
            stopDraining()
            return
        }
        emit(pending.event) { [weak self] succeeded in
            guard let self else { return }
            guard succeeded, self.store.remove(occurrenceID: pending.occurrenceID) else {
                self.stopDraining()
                return
            }
            self.emitNext()
        }
    }

    private func stopDraining() {
        lock.lock()
        isDraining = false
        lock.unlock()
    }
}
